File: core/node/bridge/adapters/sheets.js
'use strict';

/**
 * Optional reporter: push the run to a Google Sheet.
 *
 * Included to show the adapter seam rather than to be the default path. The
 * contract an adapter has to meet is the whole point:
 *
 *     name: string
 *     publish(records, summary, controls) -> object of things to print
 *
 * Nothing above this layer knows whether results end up in a file, a
 * spreadsheet, or a database, which is what makes the backend swappable
 * without touching a single test.
 *
 * Two rules this adapter follows, both learned from spreadsheets that people
 * actually maintain by hand:
 *
 *   1. Columns are located by HEADER NAME, never by fixed letter. Someone will
 *      insert a column, and every hard-coded index silently starts writing into
 *      the wrong field -- silently, because a wrong value still looks like a value.
 *
 *   2. Human-owned columns are never overwritten. Triage notes, assignees and
 *      review states belong to people; a sync that rewrites the whole row
 *      destroys work with no trace.
 */

const name = 'sheets';

const HUMAN_OWNED_COLUMNS = ['Assignee', 'Triage note', 'Review state', 'Ticket'];

function requireEnv(key) {
  const v = process.env[key];
  if (!v) {
    throw new Error(
      key + ' is not set. Copy .env.example to .env and fill it in, ' +
      'or leave REPORTER unset to use the local file reporter.'
    );
  }
  return v;
}

/**
 * Map a header row to column indices, so writes address columns by name.
 * @param {string[]} headerRow
 */
function indexColumns(headerRow) {
  const index = {};
  headerRow.forEach((title, i) => { index[String(title).trim()] = i; });
  return index;
}

/** Build the cells this adapter owns, leaving every human column untouched. */
function buildRow(record, columnIndex, existingRow) {
  const row = existingRow ? [...existingRow] : [];

  const owned = {
    'Case ID': record.caseId,
    'Module': record.module,
    'Title': record.title,
    'Status': record.status,
    'Assertions': record.counts.passed + '/' + record.counts.total,
    'Reason': record.reason || '',
    'Note': record.note || '',
    'Last run': record.ts,
  };

  for (const [header, value] of Object.entries(owned)) {
    const at = columnIndex[header];
    if (at === undefined) continue;              // column absent; do not invent one
    if (HUMAN_OWNED_COLUMNS.includes(header)) continue;
    row[at] = value;
  }
  return row;
}

async function publish(records, summary /* , controls */) {
  const spreadsheetId = requireEnv('SHEETS_SPREADSHEET_ID');
  requireEnv('SHEETS_SERVICE_ACCOUNT_JSON');

  // Wiring the Google client is left out on purpose: it would add a heavy
  // dependency to a repository whose default path needs none, and the part
  // worth reading -- header lookup, ownership split, idempotent upsert -- is
  // already above. buildRow and indexColumns are exported and unit tested.
  throw new Error(
    'The Google Sheets adapter is a documented seam, not a wired integration.\n' +
    'It would upsert ' + records.length + ' rows into spreadsheet ' + spreadsheetId + ' ' +
    '(' + summary.Passed + ' passed, ' + summary.Failed + ' failed, ' + summary.Blocked + ' blocked), ' +
    'matching rows by Case ID and writing only machine-owned columns.\n' +
    'Run without REPORTER set to use the local file reporter.'
  );
}

module.exports = { name, publish, indexColumns, buildRow, HUMAN_OWNED_COLUMNS };

File: core/python/qa_core/reporters/sheets.py
"""Optional reporter: push the run to a Google Sheet.

Included to show the adapter seam rather than to be the default path.

Two rules this reporter follows, both learned from spreadsheets that people
actually maintain by hand:

  1. Columns are located by HEADER NAME, never by fixed letter. Someone will
     insert a column, and every hard-coded index silently starts writing into
     the wrong field -- silently, because a wrong value still looks like a value.

  2. Human-owned columns are never overwritten. Triage notes, assignees and
     review states belong to people; a sync that rewrites the whole row destroys
     work with no trace.
"""

from __future__ import annotations

import os

NAME = "sheets"

HUMAN_OWNED_COLUMNS = ("Assignee", "Triage note", "Review state", "Ticket")


def require_env(key: str) -> str:
    value = os.environ.get(key)
    if not value:
        raise RuntimeError(
            f"{key} is not set. Copy .env.example to .env and fill it in, "
            "or leave REPORTER unset to use the local file reporter."
        )
    return value


def index_columns(header_row: list[str]) -> dict[str, int]:
    """Map a header row to column indices, so writes address columns by name."""
    return {str(title).strip(): i for i, title in enumerate(header_row)}


def build_row(record: dict, column_index: dict[str, int], existing_row: list | None = None) -> list:
    """Build the cells this reporter owns, leaving every human column untouched."""
    row = list(existing_row) if existing_row else []

    owned = {
        "Case ID": record["caseId"],
        "Module": record["module"],
        "Title": record["title"],
        "Status": record["status"],
        "Assertions": f'{record["counts"]["passed"]}/{record["counts"]["total"]}',
        "Reason": record.get("reason", ""),
        "Note": record.get("note", ""),
        "Last run": record["ts"],
    }

    for header, value in owned.items():
        at = column_index.get(header)
        if at is None:                          # column absent; do not invent one
            continue
        if header in HUMAN_OWNED_COLUMNS:
            continue
        while len(row) <= at:
            row.append("")
        row[at] = value

    return row


def publish(records: list[dict], summary: dict, controls: dict) -> dict:
    spreadsheet_id = require_env("SHEETS_SPREADSHEET_ID")
    require_env("SHEETS_SERVICE_ACCOUNT_JSON")

    # Wiring the Google client is left out on purpose: it would add a heavy
    # dependency to a repository whose default path needs none, and the part
    # worth reading -- header lookup, ownership split, idempotent upsert -- is
    # already above. build_row and index_columns are exported and unit tested.
    raise RuntimeError(
        "The Google Sheets reporter is a documented seam, not a wired integration.\n"
        f"It would upsert {len(records)} rows into spreadsheet {spreadsheet_id} "
        f'({summary["Passed"]} passed, {summary["Failed"]} failed, '
        f'{summary["Blocked"]} blocked), matching rows by Case ID and writing '
        "only machine-owned columns.\n"
        "Run without REPORTER set to use the local file reporter."
    )
